File: sender-android/app/src/main/java/com/openaudiolink/protocol/HandshakePayloads.kt
package com.openaudiolink.protocol

import java.io.ByteArrayOutputStream

object HandshakePayloads {
    fun hello(deviceName: String, appVersion: String): ByteArray = bytes {
        string(deviceName); string(appVersion)
        byte(ProtocolConstants.MajorVersion); byte(ProtocolConstants.MinorVersion); byte(ProtocolConstants.PlatformAndroid)
        uint32(ProtocolConstants.CapabilityAacSupported)
    }

    fun welcome(result: Int, receiverName: String, appVersion: String, sessionId: Long): ByteArray = bytes {
        byte(result); string(receiverName); string(appVersion); uint64(sessionId)
    }

    fun startStream(codec: Int, sampleRate: Long, channels: Int, bitrate: Long, frameDurationMs: Int): ByteArray = bytes {
        byte(codec); uint32(sampleRate); byte(channels); uint32(bitrate); uint16(frameDurationMs)
    }

    fun streamReady(result: Int, codec: Int, sampleRate: Long, channels: Int): ByteArray = bytes {
        byte(result); byte(codec); uint32(sampleRate); byte(channels)
    }

    fun ping(sequenceNumber: Long, timestamp: Long): ByteArray = bytes { uint32(sequenceNumber); uint64(timestamp) }
    fun error(code: Int, severity: Int, message: String): ByteArray = bytes { uint16(code); byte(severity); string(message) }

    private fun bytes(block: Writer.() -> Unit): ByteArray = Writer().apply(block).out.toByteArray()

    private class Writer {
        val out = ByteArrayOutputStream()
        fun byte(value: Int) = out.write(value)
        fun uint16(value: Int) = out.write(byteArrayOf((value ushr 8).toByte(), value.toByte()))
        fun uint32(value: Long) = out.write(ByteArray(4).also { PacketWriter.writeUInt32(it, 0, value) })
        fun uint64(value: Long) = out.write(ByteArray(8).also { PacketWriter.writeUInt64(it, 0, value) })
        fun string(value: String) {
            val data = value.toByteArray(Charsets.UTF_8)
            uint16(data.size)
            out.write(data)
        }
    }
}

File: receiver-windows/src/OpenAudioLink/Protocol/HandshakePayloads.cs
using System;
using System.Text;

namespace OpenAudioLink.Protocol
{
    public static class HandshakePayloads
    {
        public static byte[] Hello(string senderName, string senderVersion, byte platform, uint capabilities)
        {
            return Join(
                WriteString(senderName),
                WriteString(senderVersion),
                new[] { ProtocolConstants.MajorVersion, ProtocolConstants.MinorVersion, platform },
                WriteUInt32(capabilities));
        }

        public static byte[] Welcome(byte result, string receiverName, string receiverVersion, ulong sessionId)
        {
            return Join(new[] { result }, WriteString(receiverName), WriteString(receiverVersion), WriteUInt64(sessionId));
        }

        public static byte[] StartStream(byte codec, uint sampleRate, byte channels, uint bitrate, ushort frameDuration)
        {
            return Join(new[] { codec }, WriteUInt32(sampleRate), new[] { channels }, WriteUInt32(bitrate), WriteUInt16(frameDuration));
        }

        public static byte[] StreamReady(byte result, byte codec, uint sampleRate, byte channels)
        {
            return Join(new[] { result, codec }, WriteUInt32(sampleRate), new[] { channels });
        }

        public static byte[] Audio(byte codec, uint frameNumber, ulong captureTimestamp, ushort frameDuration, byte[] encodedData)
        {
            encodedData = encodedData ?? new byte[0];
            return Join(new[] { codec }, WriteUInt32(frameNumber), WriteUInt64(captureTimestamp), WriteUInt16(frameDuration), WriteUInt32((uint)encodedData.Length), encodedData);
        }

        public static byte[] Ping(uint sequence, ulong timestamp)
        {
            return Join(WriteUInt32(sequence), WriteUInt64(timestamp));
        }

        public static byte[] Error(ushort code, byte severity, string description)
        {
            return Join(WriteUInt16(code), new[] { severity }, WriteString(description));
        }

        public static byte ReadStartStreamCodec(byte[] payload)
        {
            if (payload == null || payload.Length < 12)
            {
                throw new PacketParseException("START_STREAM payload is too short.");
            }

            return payload[0];
        }

        private static byte[] WriteString(string value)
        {
            byte[] text = Encoding.UTF8.GetBytes(value ?? string.Empty);
            if (text.Length > ushort.MaxValue)
            {
                throw new ArgumentOutOfRangeException(nameof(value), "String is too long.");
            }

            return Join(WriteUInt16((ushort)text.Length), text);
        }

        private static byte[] WriteUInt16(ushort value)
        {
            byte[] buffer = new byte[2];
            PacketWriter.WriteUInt16(buffer, 0, value);
            return buffer;
        }

        private static byte[] WriteUInt32(uint value)
        {
            byte[] buffer = new byte[4];
            PacketWriter.WriteUInt32(buffer, 0, value);
            return buffer;
        }

        private static byte[] WriteUInt64(ulong value)
        {
            byte[] buffer = new byte[8];
            PacketWriter.WriteUInt64(buffer, 0, value);
            return buffer;
        }

        private static byte[] Join(params byte[][] parts)
        {
            int length = 0;
            foreach (byte[] part in parts)
            {
                length += part.Length;
            }

            byte[] output = new byte[length];
            int offset = 0;
            foreach (byte[] part in parts)
            {
                Buffer.BlockCopy(part, 0, output, offset, part.Length);
                offset += part.Length;
            }

            return output;
        }
    }
}
